tests/test_circulation_data.py

```diff
@@ -0,0 +1,265 @@
+        # Note: Mirroring tests passing does not guarantee that all code now 
+        # correctly calls on CirculationData, as well as Metadata.  This is a risk.
+
+        mirror = DummyS3Uploader()
+        # Here's a book.
+        edition, pool = self._edition(with_license_pool=True)
+
+        # Here's a link to the content of the book, which will be mirrored.
+        link_mirrored = LinkData(
+            rel=Hyperlink.OPEN_ACCESS_DOWNLOAD, href="http://example.com/",
+            media_type=Representation.EPUB_MEDIA_TYPE,
+            content="i am a tiny book"
+        )
+
+        # This link will not be mirrored.
+        link_unmirrored = LinkData(
+            rel=Hyperlink.DRM_ENCRYPTED_DOWNLOAD, href="http://example.com/2",
+            media_type=Representation.EPUB_MEDIA_TYPE,
+            content="i am a pricy book"
+        )
+
+        # Apply the metadata.
+        policy = ReplacementPolicy(mirror=mirror)
+        circulation_data = CirculationData(
+        	data_source=edition.data_source, 
+        	primary_identifier=edition.primary_identifier,
+        	links=[link_mirrored, link_unmirrored],
+        )
+        circulation_data.apply(pool, replace=policy)
+        
+
+        # Only the open-access link has been 'mirrored'.
+        # TODO: make sure the refactor is done right, and metadata does not upload
+        #eq_(0, len(mirror.uploaded))
+        [book] = mirror.uploaded
+
+        # TODO: make sure the refactor is done right, and circulation does upload 
+        #circulation = CirculationData(links=[link_mirrored, link_unmirrored], data_source=edition.data_source)
+        #circulation.apply(pool, replace=policy)
+        #[book] = mirror.uploaded
+
+        # It's remained an open-access link.
+        eq_(
+            [Hyperlink.OPEN_ACCESS_DOWNLOAD], 
+            [x.rel for x in book.resource.links]
+        )
+
+
+        # It's been 'mirrored' to the appropriate S3 bucket.
+        assert book.mirror_url.startswith('http://s3.amazonaws.com/test.content.bucket/')
+        expect = '/%s/%s.epub' % (
+            edition.primary_identifier.identifier,
+            edition.title
+        )
+        assert book.mirror_url.endswith(expect)
+
+        # make sure the mirrored link is safely on edition
+        sorted_edition_links = sorted(edition.license_pool.identifier.links, key=lambda x: x.rel)
+        unmirrored_representation, mirrored_representation = [edlink.resource.representation for edlink in sorted_edition_links]
+        assert mirrored_representation.mirror_url.startswith('http://s3.amazonaws.com/test.content.bucket/')
+
+        # make sure the unmirrored link is safely on edition
+        eq_('http://example.com/2', unmirrored_representation.url)
+        # make sure the unmirrored link has not been translated to an S3 URL
+        eq_(None, unmirrored_representation.mirror_url)
+
+
+    def test_mirror_open_access_link_fetch_failure(self):
```

Comment:
there should be a separate class that tests the mirror mixin class - it can still use circulation data as the example class